### PR TITLE
Add bzlmod support

### LINKED ISSUE #3759: Bzlmod support
## 🚀 Feature

At present your Bazel support and dependencies are managed via your WORKSPACE file. Bazel 6 introduced Bzlmod for managing project dependencies using a registry, which helps to solve the transitive dependency problem.

I am attempting to support grpc-gateway under bzlmod [here](https://github.com/rules-proto-grpc/rules_proto_grpc/tree/next), but both paths I have tried presently do not work (direct plugins and Gazelle). My third option is to maintain a patchset on top of grpc-gateway that fixes dependency loads to work under bzlmod, which really won't be pleasant. Before I do this, are you considering adding Bzlmod support? Looking at what's presently in your WORKSPACE, I don't see any major blockers.

Some useful links:
- https://bazel.build/external/migration
- https://github.com/bazelbuild/rules_go/blob/master/docs/go/core/bzlmod.md

### PR DESCRIPTION
**Pull request by AlejoAsd**
#### References to other Issues or PRs

Fixes #3759.

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md)?

Yes.

#### Brief description of what is fixed or changed

As mentioned in #3759, Bazel has introduced bzlmod to better handle dependencies across projects. In order to enable bzlmod support, the existing Bazel `WORKSPACE` must be migrated to and temporarily co-exist with a `MODULE.bazel` file, which has some similar syntax but works differently.

This PR introduces a `MODULE.bazel` file, lock and non-module dependencies file. It allows building and importing the project as a Bazel module.

This work is heavily based on the [great work of @SanjayVas](https://github.com/grpc-ecosystem/grpc-gateway/issues/3759#issuecomment-1861737758), with some dependency versions updated to match the existing `WORKSPACE` configuration.

This was tested by building the entire repository with and without bzlmod with the following commands:

```bash
# bzlmod build
bazel run //:gazelle; bazel build //...

# WORKSPACE build
bazel run //:gazelle; bazel build //... --noenable_bzlmod
```

#### Other comments

* This is a first step towards having the project be available in the Bazel registry as a module, but not enough on its own. 
* The Bazel `module` version is set to `0.0.0` which will have to be changed if this module is to be published to the Bazel registry. Once this package is in the Bazel registry, the `version` will have to be updated every time a new release is created.
* Even though this change does not register the project as a module in the registry, it can now be imported by defining a `bazel_dep` along with a suitable override in downstream projects using bzlmod.

### CODE REVIEW

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Thanks a lot for this! It seems to be failing to run `buildifier` at the moment. I'm definitely interested in having this done and migrating away from WORKSPACE over time, but we need to be sure we understand the added maintenance burden.

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:3]: (Disclaimer: I am being super verbose about how Bazel works because I've read in some issues that you don't usually work with Bazel. Please let me know if I am being too verbose!)

Ideally it should be done with every new release. The truth is that this will not matter as much until the package is available through the [Bazel registry](https://registry.bazel.build/), because right now you need to manually specify the git hash or download the archive to make this available to other projects.

As an example, in my downstream project, this is how I would end up using this:

```starlark
# The `version` here has to match the `module` version in the project's MODULE.bazel.
bazel_dep(name = "grpc_ecosystem_grpc_gateway", version = "2.23.0")
# Tell Bazel to pull directly from the repo
git_override(
    module_name = "grpc_ecosystem_grpc_gateway",
    remote = "https://github.com/grpc-ecosystem/grpc-gateway.git",
    commit = "v2.23.0"  # This could be a any version, it's not limited to tags
)
```

As long as the version I specify in `bazel_dep` matches the version in the `module` declaration in `MODULE.bazel` this will work. It does become a little more strict when publishing to the Bazel registry though.

I believe that it would make sense to only update the `module` version when releasing, to reduce the manual burden and to better prepare the workflow for when (and if) the project is published to the Bazel registry.

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:21]: That's right! Gazelle automatically handles all go dependencies by looking at the `go.mod` file and creating a dependency tree. It also handles transitive dependencies. 

The full flow is something like this:

```mermaid
flowchart LR
    go_mod[go.mod]
    gazelle[gazelle]
    buildifier[buildifier]
    bazel_mod[MODULE.bazel]
    bazel_build[BUILD.bazel]

    go_mod--Reads-->gazelle
    gazelle--Uses-->buildifier
    gazelle--Updates-->bazel_build
    buildifier--Modifies-->bazel_mod
```

An important thing to note is that the `MODULE.bazel` file isn't updated automatically when running `gazelle`. Instead, running the `gazelle` command prompts the actual dev to run `buildifier` commands by copying and pasting them.

So in order to update Bazel dependencies for go files, we can just run:

```bash
go mod tidy
bazel run //:gazelle
# Optionally, if the gazelle command requests running buildifier, copy and paste those commands in a terminal
```

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:31]: These are necessary because Gazelle only generates dependencies for Go files (since it generates dependencies by reading the `go.mod` file).

These two dependencies are introduced by `.proto` files and are not included in `go.mod`, so by adding these we tell Gazelle that those should be downloaded as well.

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:45]: That's right. This has to do with the way Bazel modules and Gazelle work that is a little bit different from working with Bazel workspaces.

This was sort of already happening in the same way with the existing WORKSPACE workflow, only the dependencies were defined in the [`repositories.bzl`](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/repositories.bzl) file and [loaded in the WORKSPACE instead](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/WORKSPACE#L117-L119).

With this new workflow and because of some changes in how Bazel modules work with respect to workspaces, we need to make these explicitly available to the project with `use_repo`.

> Will buildifier update this file if we add new dependencies to go.mod?

Yes, whenever we run `bazel run //:gazelle`, Gazelle will detect differences and prompt the dev to run `buildifier` to update this file.

File: non_module_deps.bzl
Comment (AlejoAsd) [reply non_module_deps.bzl:14]: Thanks, I didn't run `buildifier`. Let me fix this.

Comment (AlejoAsd) [reply non_module_deps.bzl]: I believe I fixed the issues. Could you try running CI again please?

Comment (AlejoAsd) [reply non_module_deps.bzl]: Looks like I was missing a newline. This should be the last fix.

File: MODULE.bazel
Comment (johanbrandhorst) [reply MODULE.bazel:3]: Thanks for clarifying, to be clear, it sounds like we'd need to update the `module` version _before_ tagging a new release tag in git? Is there any way we could punt that requirement until we decide to support the bazel registry down the road? If users have to override with git anyway, is it still a requirement for us to set it accurately?

File: MODULE.bazel
Comment (johanbrandhorst) [reply MODULE.bazel:21]: I can't believe you just drew me a mermaid diagram to explain this, that's amazing. As far as our CI checks go, as long as buildifier would detect any missed changes, I'm happy with it.

File: MODULE.bazel
Comment (johanbrandhorst) [reply MODULE.bazel:34]: I'm confused. If they are go deps, they should be in the go.mod file, since we check in our protobuf files and any dependencies they have will form part of the go.mod dependency closure at Go dependency resolution time. If these are pure protobuf dependencies (AFAIK, we don't have any proto dependencies other than googleapis), why are they part of `go_deps`?

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Looks like there's still a small generation error with the lock file - is there a risk that the contents of the lock file will change over time? I really like being able to re-run buildifier and have the output be deterministic barring changes to `go.mod` or (I guess) `MODULE.bazel`.

**Comment by johanbrandhorst** [comment]
Interesting, when I check this out and run `bazel run gazelle` locally, I see no diff, but for some reason there's a diff in CI?

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:3]: Yes, we would have to do it before tagging, which can be a bit annoying if you don't already have a release PR process defined. 

> Is there any way we could punt that requirement until we decide to support the bazel registry down the road? If users have to override with git anyway, is it still a requirement for us to set it accurately?

This comment prompted me to try to not set `version` (as it is optional) and it looks like it just works. According to the docs, not setting `version` introduces limitations when the module is depended on, but it looks like this doesn't apply if you override the dep. I am able to `bazel_dep` and `git_override` with no issues. 

An alternative approach is to use something like `version = "0.0.0"` for the time being until we introduce Bazel registry support. I think I'm going with this for the time being just because it's a little more compliant when importing the project as a module.

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:34]: Great question.

Basically, both of these repositories were used in targets that are not managed by Gazelle.

1. `google.golang.org/grpc/cmd/protoc-gen-go-grpc` is used in a [`go_proto_compiler` target](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/BUILD.bazel#L62). Gazelle does not manage `go_proto_compiler` targets.
2. `github.com/golang/protobuf` is used in a [`go_proto_library` target that compiles example protos to go](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/examples/internal/proto/examplepb/BUILD.bazel#L103-L124). While `go_proto_library` is managed by Gazelle, there are a number of [Gazelle directives at the beginning of this file](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/examples/internal/proto/examplepb/BUILD.bazel#L10-L37) that tell it to not manage the Go files compiled from these protos, which causes `go.mod` to not include the dependency.

Note that this was not an issue in the `WORKSPACE` version because we had a giant file that defined both dependencies manually ([1](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/repositories.bzl#L306-L311), [2](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/repositories.bzl#L1819-L1824)).

Unless we change the excludes, these dependencies will have to be specified manually.

**Comment by AlejoAsd** [comment]
The `@@rules_rust` hash that is causing the job to fail is some transitive dependency that is not explicitly defined in the `MODULE.bazel` file.

The only thing I've found that could affect the difference in the lockfile is the `.bazelrc` config being used in CI pipelines. Specifically, the fact that we are using a [cache](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/.github/workflows/ci.yml#L82). 

It might be the case that some dependency in the cache is causing the lockfile to turn out slightly differently. Could you try running CI for this PR without cache?

If this turns out to be the cause, a simple solution is to change [the command that is failing](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/.github/workflows/ci.yml#L90C14-L90C58) so that it does not diff the lockfile.

```bash
bazel run //:gazelle && git diff --exit-code -- !MODULE.bazel.lock
```

I don't know if this is something we want though.

File: MODULE.bazel
Comment (johanbrandhorst) [reply MODULE.bazel:36]: Thanks for the explanation, I'll categorize this in my mind as "because bazel".

File: MODULE.bazel
Comment (johanbrandhorst) [reply MODULE.bazel]: Sounds good, thanks!

**Comment by johanbrandhorst** [comment]
> The `@@rules_rust` hash that is causing the job to fail is some transitive dependency that is not explicitly defined in the `MODULE.bazel` file.
> 
> The only thing I've found that could affect the difference in the lockfile is the `.bazelrc` config being used in CI pipelines. Specifically, the fact that we are using a [cache](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/.github/workflows/ci.yml#L82).
> 
> It might be the case that some dependency in the cache is causing the lockfile to turn out slightly differently. Could you try running CI for this PR without cache?

Looking at the CI job, it doesn't look like it's actually able to match a cache in this case, so I don't think that's the reason.

> If the cache turns out to be the problem, a simple solution is to change [the command that is failing](https://github.com/grpc-ecosystem/grpc-gateway/blob/86a430974ccf379524c80a4cd89759c8f5b91d4c/.github/workflows/ci.yml#L90C14-L90C58) so that it does not diff the lockfile.
> 
> ```shell
> bazel run //:gazelle && git diff --exit-code -- !MODULE.bazel.lock
> ```
> 
> I don't know if this is something we want though.

Thanks for the suggestion, but yeah that's definitely not something we want. We'll have to figure out some way to make it deterministic, I never want to see a git diff after running the generation script. Can we specify the version to use for this transitive dependency to avoid it being resolved dynamically?

File: MODULE.bazel
Comment (AlejoAsd) [reply MODULE.bazel:36]: Haha, yeah, pretty much.

**Comment by AlejoAsd** [comment]
Looks like that pinning the `rules_rest` version did it. I'll add a comment explaining why we added that dependency and we should be good to merge after that.

**Review by johanbrandhorst — APPROVED** [review]
Thank you so much for the help and the explanations! I wonder, do we need to add anything to our README regarding how to depend on the project as a bazel user? We don't need to do it as part of this PR.

**Comment by AlejoAsd** [comment]
Excellent question. Let me take a look and update the README if necessary before merging.

**Comment by AlejoAsd** [comment]
Hmm, it looks like the only real "documentation" we have for setting up things with Bazel is this [example repo shared in an issue comment](https://github.com/grpc-ecosystem/grpc-gateway/issues/2140#issuecomment-883832999).

Since we're not publishing to the Bazel Registry just yet, I wanted to add a short user guide on how to import this repo as a Bazel module, but I don't think there is a README or guide that goes into Bazel integration.

Let's follow up on another PR so that we can at least merge this. Thanks a ton for the thorough reviews Johan!

**Comment by johanbrandhorst** [comment]
Thanks for your contribution! If you see any other QoL changes we can make for Bazel users, please don't hestitate to contribute :)

**Comment by ns-ggeorgiev** [comment]
I tried to adopt your repo through bzlmod but it complains about missing `com_github_bazelbuild_buildtools`

**Comment by AlejoAsd** [comment]
Thanks @ns-ggeorgiev, I have opened https://github.com/grpc-ecosystem/grpc-gateway/pull/4974 to address this.

**Pull request merged**
